In [3]:
import numpy as np

import pandas as pd

from sklearn.linear_model import LogisticRegression  # not used yet!

import statsmodels.discrete.discrete_model as sm


## Train a logistic model on the test set

Here we want to train a logistic model and choose features based on the p-values of the coefficients

We need to do a little bit of data cleaning and removal of rows containing missing data as seen in the data exploration



In [4]:
df = pd.read_csv('../data/processed/cs-training.csv')

df = df[np.isfinite(df['MonthlyIncome'])]

df = df[np.isfinite(df['NumberOfDependents'])]

df = df[~df.isin([np.nan, np.inf, -np.inf]).any(1)]

X = df.loc[:, df.columns != 'SeriousDlqin2yrs']

del X['Unnamed: 0']

y = df['SeriousDlqin2yrs']


logit_model=sm.Logit(y,X)
result=logit_model.fit()
print(result.summary2())

Optimization terminated successfully.
         Current function value: 0.237361
         Iterations 7
                                   Results: Logit
Model:                      Logit                  Pseudo R-squared:       0.059     
Dependent Variable:         SeriousDlqin2yrs       AIC:                    57114.2447
Date:                       2019-11-18 16:14       BIC:                    57211.2196
No. Observations:           120269                 Log-Likelihood:         -28547.   
Df Model:                   9                      LL-Null:                -30345.   
Df Residuals:               120259                 LLR p-value:            0.0000    
Converged:                  1.0000                 Scale:                  1.0000    
No. Iterations:             7.0000                                                   
-------------------------------------------------------------------------------------
                                      Coef.  Std.Err.    z     P>|z|   [0.

We can see that all 10 predictors have a p-value less than 0.05 except for the RevolvingUtil... 
so we could consider them significant.

An interesting thing to note about this result is that the pseudo R^2 is 0.059. Not good!!

This tells us that the model including features at the 1st order is not really working!! We need some more interactions, or higher orders in the predictors to improve the model!!

I kind ran out of time at this point

things to do: 

1. include interactions and higher order features

2. use clustering apprach (e.g. k-means etc.)

3. use a neural network approach